### PR TITLE
Fix region propagation of finally end blocks

### PR DESCRIPTION
**Pull request by meg-gupta**
We add edge from infinite loops in finally region to end of finally block, so that unreachable block elimination does not remove finally end block. We have to retain finally end block, so that we can add edge from finally end block to early exit. 

When the only edge to finally block is an edge from a loop within some child region, we can end up in propagating incorrect region info from the predecessor. We have to use the region on the finally label for such an end of finally block. 

So create a new map called leaveNullLabelToFinallyLabelMap and pick up region info from here.

Also piggybacking other small changes:
- dumping
- fix asserts for Br instructions : 
Leaves in non exception finally are converted to Br instructions, for such branches the target region is the parent of the current region.
Also, a BrOnException from finally to early exit can be converted to BrOnNoException and Br, we cannot assert for such a Br that the target region is same as Br region

### CODE REVIEW

**Comment by meg-gupta** [comment]
It is easy to review this commit by commit. This PR is stacked on top of https://github.com/Microsoft/ChakraCore/pull/3090. 

The only commits to review are  on top of "Fix region asserts for Leaves converted to Br in non exception finally"

**Comment by meg-gupta** [comment]
@dotnet-bot Please test Ubuntu _no_jit_shared_ubuntu_linux_test

File: lib/Backend/IR.h [hunk 1]
```diff
@@ -748,10 +750,11 @@ class BranchInstr : public Instr
     static BranchInstr * New(Js::OpCode opcode, Opnd* destOpnd, LabelInstr * branchTarget, Opnd *srcOpnd, Func *func);
     static BranchInstr * New(Js::OpCode opcode, LabelInstr * branchTarget, Opnd *src1Opnd, Opnd *src2Opnd, Func *func);
 
-    BranchInstr(bool hasBailOutInfo = false) : Instr(hasBailOutInfo), m_branchTarget(nullptr), m_isAirlock(false), m_isSwitchBr(false), m_isOrphanedLeave(false), m_areCmpRegisterFlagsUsedLater(false)
+    BranchInstr(bool hasBailOutInfo = false) : Instr(hasBailOutInfo), m_branchTarget(nullptr), m_isAirlock(false), m_isSwitchBr(false), m_isOrphanedLeave(false), m_areCmpRegisterFlagsUsedLater(false), m_brFinallyToEarlyExit(nullptr)
```
Comment (rajatd) [inline lib/Backend/IR.h]: >nullptr [](start = 224, length = 7)

initialize to false

File: lib/Backend/FlowGraph.cpp [hunk 2]
```diff
@@ -1721,21 +1767,26 @@ FlowGraph::Destroy(void)
                         Assert(!this->func->DoGlobOpt());
                         break;
                     case Js::OpCode::BrOnNoException:
-                        Assert(region->GetType() == RegionTypeTry || region->GetType() == RegionTypeCatch || region->GetType() == RegionTypeFinally);
+                        Assert(region->GetType() == RegionTypeTry || region->GetType() == RegionTypeCatch || region->GetType() == RegionTypeFinally ||
+                            (predRegion->GetType() == RegionTypeFinally && predBlock->GetLastInstr()->AsBranchInstr()->m_brFinallyToEarlyExit));
```
Comment (rajatd) [inline lib/Backend/FlowGraph.cpp]: >predRegion->GetType() == RegionTypeFinally && predBlock->GetLastInstr()->AsBranchInstr()->m_brFinallyToEarlyExit)) [](start = 29, length = 114)

This condition will be true when the early exit code was moved out as a break block?

File: lib/Backend/FlowGraph.cpp [hunk 3]
```diff
@@ -636,6 +673,13 @@ void FlowGraph::InsertEdgeFromFinallyToEarlyExit(BasicBlock *finallyEndBlock, IR
     IR::LabelInstr *leaveLabel = IR::LabelInstr::New(Js::OpCode::Label, this->func);
     lastInstr->InsertBefore(leaveLabel);
 
+    if (this->leaveNullLabelToFinallyLabelMap->ContainsKey(finallyEndBlock->GetFirstInstr()->AsLabelInstr()))
+    {
+        // Add new LeaveNull label to the map, we don't delete old LeaveNull label from the map
+        Assert(this->leaveNullLabelToFinallyLabelMap->ContainsKey(finallyEndBlock->GetFirstInstr()->AsLabelInstr()));
```
Comment (rajatd) [inline lib/Backend/FlowGraph.cpp]: >this->leaveNullLabelToFinallyLabelMap->ContainsKey(finallyEndBlock->GetFirstInstr()->AsLabelInstr()) [](start = 15, length = 100)

Looks like the same condition you checked in the above `if`

Comment (meg-gupta) [inline lib/Backend/FlowGraph.cpp]: Fixed.

File: lib/Backend/IR.h (hunk 1 above)
Comment (meg-gupta) [inline lib/Backend/IR.h]: Fixed.

File: lib/Backend/FlowGraph.cpp (hunk 2 above)
Comment (meg-gupta) [inline lib/Backend/FlowGraph.cpp]: Added comment and test case for this. This happens when the early exit code is moved inside the finally block. And the early exit has 2 edges one from a root region and one from the preceeding BrOnNoException block.

**Comment by LouisLaf** [comment]
Now that I look back at this:
"We add edge from infinite loops in finally region to end of finally block, so that unreachable block elimination does not remove finally end block. We have to retain finally end block, so that we can add edge from finally end block to early exit."

Is there a way to have your code work without the edge?  If the finally block has an infinite loop, the code after the early exit isn't reachable...

**Comment by meg-gupta** [comment]
@LouisLaf I don't suspect globopt to run into problems without this edge. But one issue that comes up with not adding these edges is in building loops/processing basic blocks in flowgraphs. Mostly due to the fact that we don't cascade edge removal in unreachable block elimination.

**Comment by rajatd** [comment]
Looks good to me as it is now, barring @LouisLaf 's question.

**Comment by meg-gupta** [comment]
@LouisLaf First look without the edge, seems a like a few FlowGraph methods need changes. Ok if I visit this issue in a different PR ? Can I merge this one ?

**Comment by meg-gupta** [comment]
Spoke to @LouisLaf offline. Will be merging this.

**Pull request merged**